### PR TITLE
Don't use FIRST_CUDA on generators which don't support toolsets

### PR DESCRIPTION
**Pull request by larroy**
## Description ##
Unix Makefiles don't support toolsets, only Visual Studio and Xcode.
This fixes build for newer CMake versions on other platforms.

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 3.0.2)
-
-if((${CMAKE_VERSION} VERSION_GREATER "3.9.0") OR (${CMAKE_VERSION} VERSION_EQUAL "3.9.0"))
+message(STATUS "CMake version '${CMAKE_VERSION}' using generator '${CMAKE_GENERATOR}'")
+if(((${CMAKE_GENERATOR} MATCHES "Visual Studio.*") OR (${CMAKE_GENERATOR} MATCHES "Xcode.*"))
```
Comment (cjolivier01) [inline CMakeLists.txt]: You left out "Unix Makefiles" and broke the Linux builds, it seems.

Comment (larroy) [reply CMakeLists.txt]: Makefiles don't support generators, that's whe whole point

Comment (larroy) [reply CMakeLists.txt]: Your latest commit broke it again for me. It also passed CI, can we find out what combination of cmake version / platform is not working?

Comment (larroy) [reply CMakeLists.txt]: Hi Chris. I tested cmake 3.5.1 in Linux using Makefiles and Ninja and cmake 3.9.6 under CLion using Make and both work with my commit.

**Comment by cjolivier01** [comment]
simply building isn’t what was broken — cmake knowing how to build cuda
files natively was broken by your change, which increases build time
significantly.
